plugins/api-docs/src/components/GraphQlDefinitionWidget/GraphQlDefinitionWidget.tsx

```diff
@@ -44,21 +45,50 @@ const useStyles = makeStyles<BackstageTheme>(() => ({
 
 type Props = {
   definition: any;
+  graphqlLink: any;
 };
 
-export const GraphQlDefinitionWidget = ({ definition }: Props) => {
+export const GraphQlDefinitionWidget = ({ definition, graphqlLink }: Props) => {
```

Comment:
My suggestion here would be that we add a `fetcher` prop that defaults to the existing read-only one. That way it will be possible to return a specialised `GraphQlDefinitionWidget` from the custom API rendering definition, which closes over the `apiEntity` on order to grab information. For example this:

```ts
createApiFactory({
  api: apiDocsConfigRef,
  deps: {},
  factory: () => {
    const definitionWidgets = defaultDefinitionWidgets();
    return {
      getApiDefinitionWidget: (apiEntity: ApiEntity) => {
        if (apiEntity.spec.type === 'graphql') {
          const fetcher = async graphQLParams => {
            const endpoint = entity.metadata.annotations?.['any-graphql-endpoint-annotation']
            const res = fetch(endpoint, { ... })
            return res.json()
          }

          return {
            type: 'graphql',
            title: 'GraphQL',
            component: definition =>
              <GraphQlDefinitionWidget definition={definition} fetcher={fetcher} />,
          } as ApiDefinitionWidget;
        }

        // fallback to the defaults
        return definitionWidgets.find(d => d.type === apiEntity.spec.type);
      },
    };
  },
})
```

Reply:
I agree. My main issues was around the inflexibility of the annotation approach for general use. Having a code based approach with small API seems like a good middle ground in the short term.